idaes/core/util/model_diagnostics.py

```diff
@@ -0,0 +1,611 @@
+# at the URL "https://github.com/IDAES/idaes-pse".
+##############################################################################
+"""Degeneracy Hunter is a collection of utility functions to assist in mathematical
+modeling in Pyomo.
+"""
+
+__author__ = "Alexander Dowling"
+
+from pyomo.environ import SolverFactory, Var, Param, Constraint, Objective, ConcreteModel, Set, Binary
+from pyomo.core.expr.visitor import identify_variables
+from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
+import numpy as np
+from scipy.sparse.linalg import svds
+from scipy.sparse import issparse, find
+
+from idaes.core.util.model_statistics import large_residuals_set, variables_near_bounds_set
+
+from pyomo.opt import SolverStatus, TerminationCondition
+
+import matplotlib.pyplot as plt
+
+from operator import itemgetter
+
+
+class DegeneracyHunter():
+
+    def __init__(self,block_or_jac,solver=None):
+        ''' Initialize Degeneracy Hunter Object
+    
+        Arguments:
+            block_or_jac: Pyomo model or Jacobian
+            solver: Pyomo SolverFactory
+            
+        Notes:
+            Passing a Jacobian to Degeneracy Hunter is current untested.
+        
+        '''
+        
+        if type(block_or_jac) is ConcreteModel:
+        
+            # Add Pyomo model to the object
+            self.block = block_or_jac
+        
+            # setup pynumero interface
+            self.nlp = PyomoNLP(self.block)
+
+            # calculate Jacobian of equality constraints in COO sparse matrix format
+            jac_eq = self.nlp.evaluate_jacobian_eq()
+
+            # save the Jacobian
+            self.jac_eq = jac_eq
+        
+            # Create a list of equality constraint names            
+            self.eq_con_list = PyomoNLP.get_pyomo_equality_constraints(self.nlp)
+        
+            self.candidate_eqns = None
+        
+        elif type(block_or_jac) is np.array:
+        
+            raise NotImplementedError
+        
+            # TODO: Need to refactor, document, and test support for Jacobian
+            self.jac_eq = block_or_jac
+            
+            self.eq_con_list = None
+        
+        else:
+            
+            raise TypeError
+        
+        # number of equality constraints, variables
+        self.n_eq = self.jac_eq.shape[0]
+        self.n_var = self.jac_eq.shape[1]
+        
+        
+        # Define default candidate equations (enumerate)
+        candidate_eqns = range(self.n_eq)
+                
+        # Initialize solver
+        if solver is None:
+            # TODO: Test performance with open solvers such as cbc
+            self.solver = SolverFactory('gurobi')
+            self.solver.options = {'NumericFocus':3}
+            
+        else:
+            # TODO: Make this a custom exception following IDAES standards
+            assert type(solver) is SolverFactory, "Argument solver should be type SolverFactory" 
```

Comment:
I don't think the class type is correct for the `solver` argument. I think the `SolverFactory` returns a solver object which isn't a subclass of `SolverFactory` so this assert statement will never be true. @jsiirola is there a solver base class that could be checked for here?

Reply:
I have used `SystemCallSolver` in the past, but at some point changed to the opinion that a solver is anything that implements a `solve` method (probably should have an `options` data structure as well).

Reply:
I commented this out to be safe.